### PR TITLE
Generate code for unused const- and inline-fns if -Clink-dead-code is specified.

### PR DESCRIPTION
**Pull request by michaelwoerister**
Fixes https://github.com/rust-lang/rust/issues/46467.

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+

Nice!

File: src/test/codegen/link-dead-code.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// compile-flags:-Clink-dead-code
+
+#![feature(const_fn)]
+#![crate_type = "rlib"]
+
+// This test makes sure that, when -Clink-dead-code is specified, we generate
+// code for functions that would otherwise be skipped.
+
+// CHECK-LABEL: define hidden i64 @_ZN14link_dead_code8const_fn
+const fn const_fn() -> usize { 1 }
```
Comment (kennytm) [inline src/test/codegen/link-dead-code.rs]: Please change the `usize` to some more portable type. This test is failing on 32-bit platforms.

```
[00:52:46] ---- [codegen] codegen/link-dead-code.rs stdout ----
[00:52:46] 	
[00:52:46] error: verification with 'FileCheck' failed
[00:52:46] status: exit code: 1
[00:52:46] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/link-dead-code.ll" "/checkout/src/test/codegen/link-dead-code.rs"
[00:52:46] stdout:
[00:52:46] ------------------------------------------
[00:52:46] 
[00:52:46] ------------------------------------------
[00:52:46] stderr:
[00:52:46] ------------------------------------------
[00:52:46] /checkout/src/test/codegen/link-dead-code.rs:19:17: error: expected string not found in input
[00:52:46] // CHECK-LABEL: define hidden i64 @_ZN14link_dead_code8const_fn
[00:52:46]                 ^
[00:52:46] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/link-dead-code.ll:1:1: note: scanning from here
[00:52:46] ; ModuleID = 'link_dead_code0-8787f43e282added376259c1adb08b80.rs'
[00:52:46] ^
[00:52:46] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/link-dead-code.ll:8:1: note: possible intended match here
[00:52:46] define hidden i32 @_ZN14link_dead_code8const_fn17h307759a84b38b8b4E() unnamed_addr #0 {
[00:52:46] ^
[00:52:46] 
[00:52:46] ------------------------------------------
[00:52:46] 
[00:52:46] thread '[codegen] codegen/link-dead-code.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2776:8
[00:52:46] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Review by kennytm — CHANGES_REQUESTED** [review]

**Comment by pnkfelix** [comment]
@michaelwoerister are you going to be able to address the problem pointed out by kennytm?

Times getting very tight to backport this into beta...

**Comment by michaelwoerister** [comment]
Whoops, sorry, didn't see this because of PTO.

**Review by kennytm — APPROVED** [review]

**Comment by alexcrichton** [comment]
Accepting for backport as it fixes a regression https://github.com/rust-lang/rust/issues/46467

**Pull request merged**